content/courses/advanced-ada/parts/control_flow/exceptions.rst

```diff
@@ -3,19 +3,209 @@ Exceptions
+  application using checks at runtime;
+
+- bounded errors;
+
+- erroneous execution.
+
+In this section, we discuss each of these categories.
+
+.. admonition:: In the Ada Reference Manual
+
+    - :arm:`1.1.5 Classification of Errors <1-1-5>`
+
+
+.. _Adv_Ada_Compilation_Errors:
+
+Compilation errors
+~~~~~~~~~~~~~~~~~~
+
+In the category of compilation errors, the goal is to prevent compilers to
+accept illegal programs. Here, any program that doesn't follow the rules
+described in the Ada Reference Manual is considered illegal. Those rules
+include not only simple syntax errors, but also more complicated rules, such as
+the ones concerning
+:ref:`accessibility levels <Adv_Ada_Accessibility_Levels_Intro>` for access
+types.
+
+Note that Ada |mdash| in contrast to many programming languages, which can be
+quite permissive |mdash| tries to prevent as many errors as possible at
+compilation time because of its focus on safety. However, even though a wide
+range of errors can be detected at compilation time, this doesn't mean that a
+legal Ada program is free of errors. Therefore, using methods such as static
+analysis or unit testing is important.
+
+
+.. _Adv_Ada_Runtime_Errors:
+
+Runtime errors
+~~~~~~~~~~~~~~
+
+When a rule cannot be verified at compilation time, a common strategy is to
+have the compiler insert runtime checks into the resulting application. We see
+details about these checks later on when we discuss
+:ref:`checks and exceptions <Adv_Ada_Checks_And_Exceptions>`.
+
+A typical example is an :ref:`overflow check <Adv_Ada_Overflow_Check>`.
+Consider a calculation using variables: if this calculation leads to a result
+that isn't representable with the underlying data types, we cannot possibly
+store a value  into a register or memory that can be considered correct |mdash|
+so we have to detect this situation. Unfortunately, because we're using
+variables, we obviously cannot verify the result of he calculation at
+compilation time, so we have to verify it at runtime.
+
+As we've mentioned before, Ada strives for detecting as many erroneous
+conditions as possible, while other programming language would allow errors
+such as overflow errors to remain undetected |mdash| which would likely lead
+the application to misbehave. Those checks raise an exception if an erroneous
+condition is detected, so the programmer has the means |mdash| and the
+responsibility |mdash| to catch that exception and handle the situation
+properly (Note, however, that some of the runtime checks can be deactivated.
+We will discuss this topic later on.)
+
+
+.. _Adv_Ada_Bounded_Errors:
+
+Bounded errors
+~~~~~~~~~~~~~~
+
+For certain kinds of errors, the compiler might not be able to detect the error
+|mdash| neither at compilation time, nor with checks at runtime. Such errors
+are called bounded errors if their possible effects are *bounded*. In fact, the
+Ada Reference Manual describes each bounded error and its possible effects
```

Comment:
I believe they are called bounded errors *because* their possible effects are bounded.

Reply:
Yes, good point! (I've just replaced "if" by "because" in the text.)